### ISSUE TITLE
No learning occurs with images of different aspect ratios when using xformers

### ISSUE THREAD

**Issue by Zyin055** [bmaltais/kohya_ss#1116]
I updated from feb6728 (May 9) to df1a9ac (June 26) and I am no longer able to do trainings with images with different aspect ratios when using xformers (`Enable buckets` and `Don't upscale bucket resolution` are checked). No learning occurs, no errors pop up in the console, and the LoRA does not affect the output in any way during image generation. If I crop my images to all be the same aspect ratio, for example 4:5, then the training works as expected. Before updating I was able to use different aspect ratios successfully with xformers enabled, so something happened during updating that caused this to break.

I have done a bunch of testing to try and figure out the root cause. I tried different optimizers (DAdaptation, DAdaptAdam, AdamW8bit), different xformer versions (0.0.14.dev, 0.0.20), and different training settings (default, my own settings). Eventually I tried turning off xformers (`Use xformers` unchecked) and figured out this was the cause, which prompted me to post this issue.

Here is an example console output for a failed training (xformers on):
```
←[1;33m=============================================================
Modules installed outside the virtual environment were found.
This can cause issues. Please review the installed modules.

You can uninstall all local modules with:

←[1;34mdeactivate
pip freeze > uninstall.txt
pip uninstall -y -r uninstall.txt
←[1;33m=============================================================←[0m

11:00:37-915043 INFO     Version: v21.7.16
11:00:37-926015 INFO     nVidia toolkit detected
11:00:58-046457 INFO     Torch 2.0.1+cu118
11:00:58-077374 INFO     Torch backend: nVidia CUDA 11.8 cuDNN 8700
11:00:58-081364 INFO     Torch detected GPU: NVIDIA GeForce RTX 3060 VRAM 12287 Arch (8, 6) Cores 28
11:00:58-083358 INFO     Verifying modules instalation status from requirements_windows_torch2.txt...
11:00:58-092335 INFO     Verifying modules instalation status from requirements.txt...
11:01:05-499539 INFO     headless: False
11:01:05-509513 INFO     Load CSS...
Running on local URL:  http://127.0.0.1:7861

To create a public link, set `share=True` in `launch()`.
11:01:57-640648 INFO     Loading config...
11:02:26-337396 INFO     Loading config...
11:02:26-455081 INFO     LoRA type changed...
11:03:02-982646 INFO     Start training LoRA LyCORIS/LoHa ...
11:03:02-984641 INFO     Valid image folder names found in: C:\Stuff\AI\kohya_ss training\lora\test\images
11:03:02-986636 INFO     Folder 1_woman: 193 images found
11:03:02-987633 INFO     Folder 1_woman: 193 steps
11:03:02-988630 INFO     Total steps: 193
11:03:02-990625 INFO     Train batch size: 22
11:03:02-992620 INFO     Gradient accumulation steps: 1
11:03:02-993617 INFO     Epoch: 4
11:03:02-994614 INFO     Regulatization factor: 1
11:03:02-995612 INFO     max_train_steps (193 / 22 / 1 * 4 * 1) = 36
11:03:02-997606 INFO     stop_text_encoder_training = 0
11:03:02-999600 INFO     lr_warmup_steps = 0
11:03:14-901575 INFO     accelerate launch --num_cpu_threads_per_process=2 "train_network.py" --enable_bucket
                         --pretrained_model_name_or_path="C:/Stuff/AI/Stable
                         Diffusion/models/Stable-diffusion/sd-v1-5.ckpt" --train_data_dir="C:\Stuff\AI\kohya_ss
                         training\lora\test\images" --resolution="512,640" --output_dir="C:\Stuff\AI\kohya_ss
                         training\lora\test\model" --logging_dir="C:\Stuff\AI\kohya_ss
                         training\lora\test\log" --network_alpha="8" --save_model_as=safetensors
                         --network_module=lycoris.kohya --network_args "conv_dim=4" "conv_alpha=4" "use_cp=False"
                         "algo=loha" --network_dropout="0" --network_dim=8 --output_name="test11-freshinstall-restarted"                         --lr_scheduler_num_cycles="4" --learning_rate="1.0" --lr_scheduler="constant"
                         --train_batch_size="22" --max_train_steps="36" --save_every_n_epochs="1"
                         --mixed_precision="fp16" --save_precision="fp16" --seed="1234" --caption_extension=".txt"
                         --cache_latents --optimizer_type="DAdaptation" --optimizer_args --optimizer_args
                         "decouple=True" "weight_decay=0.01" "betas=0.9,0.99" --max_data_loader_n_workers="1"
                         --max_token_length=150 --bucket_reso_steps=64 --min_snr_gamma=3 --gradient_checkpointing
                         --xformers --persistent_data_loader_workers --bucket_no_upscale --wandb_api_key="False"
                         --sample_sampler=euler_a --sample_prompts="C:\Stuff\AI\kohya_ss
                         training\lora\test\model\sample\prompt.txt" --sample_every_n_epochs="1"
A matching Triton is not available, some optimizations will not be enabled.
Error caught was: No module named 'triton'
prepare tokenizer
update token length: 150
Using DreamBooth method.
prepare images.
found directory C:\Stuff\AI\kohya_ss training\lora\test\images\1_woman contains 193 image files
193 train images with repeating.
0 reg images.
no regularization images / 正則化画像が見つかりませんでした
[Dataset 0]
  batch_size: 22
  resolution: (512, 640)
  enable_bucket: True
  min_bucket_reso: 256
  max_bucket_reso: 1024
  bucket_reso_steps: 64
  bucket_no_upscale: True

  [Subset 0 of Dataset 0]
    image_dir: "C:\Stuff\AI\kohya_ss training\lora\test\images\1_woman"
    image_count: 193
    num_repeats: 1
    shuffle_caption: False
    keep_tokens: 0
    caption_dropout_rate: 0.0
    caption_dropout_every_n_epoches: 0
    caption_tag_dropout_rate: 0.0
    color_aug: False
    flip_aug: False
    face_crop_aug_range: None
    random_crop: False
    token_warmup_min: 1,
    token_warmup_step: 0,
    is_reg: False
    class_tokens: woman
    caption_extension: .txt


[Dataset 0]
loading image sizes.
100%|██████████████████████████████████████████████████████████████████████████████| 193/193 [00:00<00:00, 1244.80it/s]
make buckets
min_bucket_reso and max_bucket_reso are ignored if bucket_no_upscale is set, because bucket reso is defined by image size automatically / bucket_no_upscaleが指定された場合は、bucketの解像度は画像サイズから自動計算されるため、min_bucket_resoとmax_bucket_resoは無視されます
number of images (including repeats) / 各bucketの画像枚数（繰り返し回数を含む）
bucket 0: resolution (384, 704), count: 42
bucket 1: resolution (448, 576), count: 47
bucket 2: resolution (512, 512), count: 8
bucket 3: resolution (512, 576), count: 2
bucket 4: resolution (512, 640), count: 94
mean ar error (without repeats): 0.010828044041436207
preparing accelerator
C:\Stuff\AI\kohya_ss\venv\lib\site-packages\accelerate\accelerator.py:258: FutureWarning: `logging_dir` is deprecated and will be removed in version 0.18.0 of 🤗 Accelerate. Use `project_dir` instead.
  warnings.warn(
Using accelerator 0.15.0 or above.
loading model for process 0/1
load StableDiffusion checkpoint: C:/Stuff/AI/Stable Diffusion/models/Stable-diffusion/sd-v1-5.ckpt
loading u-net: <All keys matched successfully>
loading vae: <All keys matched successfully>
loading text encoder: <All keys matched successfully>
CrossAttention.forward has been replaced to enable xformers.
import network module: lycoris.kohya
[Dataset 0]
caching latents.
100%|████████████████████████████████████████████████████████████████████████████████| 193/193 [00:42<00:00,  4.54it/s]
Using rank adaptation algo: loha
Apply different lora dim for conv layer
Conv Dim: 4, Linear Dim: 8
Apply different alpha value for conv layer
Conv alpha: 4.0, Linear alpha: 8.0
Use Dropout value: 0.0
Create LyCORIS Module
create LyCORIS for Text Encoder: 72 modules.
Create LyCORIS Module
create LyCORIS for U-Net: 278 modules.
enable LyCORIS for text encoder
enable LyCORIS for U-Net
preparing optimizer, data loader etc.
Deprecated: use prepare_optimizer_params(text_encoder_lr, unet_lr, learning_rate) instead of prepare_optimizer_params(text_encoder_lr, unet_lr)
use D-Adaptation AdamPreprint optimizer | {'decouple': True, 'weight_decay': 0.01, 'betas': (0.9, 0.99)}
Using decoupled weight decay
running training / 学習開始
  num train images * repeats / 学習画像の数×繰り返し回数: 193
  num reg images / 正則化画像の数: 0
  num batches per epoch / 1epochのバッチ数: 12
  num epochs / epoch数: 3
  batch size per device / バッチサイズ: 22
  gradient accumulation steps / 勾配を合計するステップ数 = 1
  total optimization steps / 学習ステップ数: 36
steps:   0%|                                                                                    | 0/36 [00:00<?, ?it/s]
epoch 1/3
A matching Triton is not available, some optimizations will not be enabled.
Error caught was: No module named 'triton'
C:\Stuff\AI\kohya_ss\venv\lib\site-packages\torch\utils\checkpoint.py:31: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")
C:\Stuff\AI\kohya_ss\venv\lib\site-packages\xformers\ops\fmha\flash.py:339: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()  and inp.query.storage().data_ptr() == inp.key.storage().data_ptr()
steps:  33%|████████████████████▋                                         | 12/36 [01:14<02:28,  6.19s/it, loss=0.0673]
saving checkpoint: C:\Stuff\AI\kohya_ss training\lora\test\model\test11-freshinstall-restarted-000001.safetensors
generating sample images at step / サンプル画像生成 ステップ: 12
prompt: photo of a woman
negative_prompt: None
height: 512
width: 512
sample_steps: 30
scale: 7.5
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [00:06<00:00,  4.65it/s]
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [00:06<00:00,  4.72it/s]
epoch 2/3
steps:  67%|█████████████████████████████████████████▎                    | 24/36 [02:25<01:12,  6.08s/it, loss=0.0742]
saving checkpoint: C:\Stuff\AI\kohya_ss training\lora\test\model\test11-freshinstall-restarted-000002.safetensors
generating sample images at step / サンプル画像生成 ステップ: 24
prompt: photo of a woman
negative_prompt: None
height: 512
width: 512
sample_steps: 30
scale: 7.5
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [00:06<00:00,  4.70it/s]
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [00:06<00:00,  4.70it/s]
epoch 3/3
steps: 100%|██████████████████████████████████████████████████████████████| 36/36 [03:42<00:00,  6.17s/it, loss=0.0687]
generating sample images at step / サンプル画像生成 ステップ: 36
prompt: photo of a woman
negative_prompt: None
height: 512
width: 512
sample_steps: 30
scale: 7.5
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [00:06<00:00,  4.70it/s]
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [00:06<00:00,  4.73it/s]
saving checkpoint: C:\Stuff\AI\kohya_ss training\lora\test\model\test11-freshinstall-restarted.safetensors
model saved.
steps: 100%|██████████████████████████████████████████████████████████████| 36/36 [03:50<00:00,  6.40s/it, loss=0.0687]
```

**Comment by bmaltais** [comment]
Not sure it is related to the GUI. I train exclusively with bucketing and xformers and training is working as expected.

**Comment by ytn2001** [comment]
me too, No learning occurs

**Comment by CoreyCorza** [comment]
Me too. Same problem

**Comment by nanonomad** [comment]
Similar issue with bucketing here. Not sure if its _no learning_, but was _very_ inefficient using bucketing with off-aspect-ratio images. Cropping images to bucket sizes gets around the problem, but sort of defeats the purpose. Not sure if its padding out buckets with a handful of samples.
